modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/SnapshotCheckDistributedProcess.java

```diff
@@ -0,0 +1,549 @@
+
+        phase2PartsHashes.start(procId, locReq);
+
+        if (log.isDebugEnabled())
+            log.debug("Started partitions validation as part of the snapshot checking. Request=" + locReq + '.');
+    }
+
+    /** Finds current snapshot name from the metas. */
+    private @Nullable String snpName(@Nullable Map<UUID, ? extends List<SnapshotMetadata>> metas) {
+        if (F.isEmpty(metas))
+            return null;
+
+        // Ensure the same snapshot name or empty.
+        assert F.flatCollections(metas.values().stream().filter(Objects::nonNull).collect(Collectors.toList()))
+            .stream().map(SnapshotMetadata::snapshotName).collect(Collectors.toSet()).size() < 2
+            : "Empty or not unique snapshot names in the snapshot metadatas.";
+
+        for (List<SnapshotMetadata> nodeMetas : metas.values()) {
+            if (F.isEmpty(nodeMetas))
+                continue;
+
+            assert nodeMetas.get(0) != null : "Empty snapshot metadata in the results";
+            assert !F.isEmpty(nodeMetas.get(0).snapshotName()) : "Empty snapshot name in a snapshot metadata.";
+
+            return nodeMetas.get(0).snapshotName();
+        }
+
+        return null;
+    }
+
+    /** Starts the snapshot full validation. */
+    public IgniteInternalFuture<SnapshotPartitionsVerifyTaskResult> start(
+        String snpName,
+        @Nullable String snpPath,
+        @Nullable Collection<String> grpNames,
+        boolean inclCstHndlrs
+    ) {
+        assert !F.isEmpty(snpName);
+
+        UUID rqId = UUID.randomUUID();
+
+        GridFutureAdapter<SnapshotPartitionsVerifyTaskResult> clusterOpFut = new GridFutureAdapter<>();
+
+        clusterOpFut.listen(fut -> clusterOpFuts.remove(rqId));
+
+        stopFutureOnAnyFailure(clusterOpFut, () -> {
+            List<UUID> requiredNodes = new ArrayList<>(F.viewReadOnly(kctx.discovery().discoCache().aliveBaselineNodes(), F.node2id()));
+
+            SnapshotCheckProcessRequest req = new SnapshotCheckProcessRequest(
+                rqId,
+                kctx.localNodeId(),
+                requiredNodes,
+                requiredNodes.get(ThreadLocalRandom.current().nextInt(requiredNodes.size())),
+                snpName,
+                snpPath,
+                grpNames,
+                0,
+                inclCstHndlrs,
+                true
+            );
+
+            phase1CheckMetas.start(req.requestId(), req);
+
+            clusterOpFuts.put(rqId, clusterOpFut);
+        });
+
+        return clusterOpFut;
+    }
+
+    /**
+     * Ensures thta the future is stopped if any failure occures.
+     *
+     * @param fut Future to stop. If {@code null}, ignored.
+     * @param action Related action to launch and watch.
+     */
+    private static void stopFutureOnAnyFailure(@Nullable GridFutureAdapter<?> fut, Runnable action) {
+        if (fut == null) {
+            action.run();
+
+            return;
+        }
+
+        try {
+            action.run();
```

Comment:
Code duplication.

```
try {
    action.run();
}
catch (Throwable th) {
    if (fut != null)
        fut.onDone(th);
}
```

Reply:
Also, it looks like that the `fut` is never `null`

Reply:
Irrelevant for now. The code was changed. The null check is above. The cluster opearation result future can be null in `clean(...) for example.`